### PR TITLE
Implement vif extensions

### PR DESCRIPTION
**Pull request by fredrik-sk**
Add all codes for the vif extensions.

### CODE REVIEW

**Comment by fredrik-sk** [comment]
I don't know why it has been closed. I tried to push a new revision (a simple rebase).

**Comment by fredrik-sk** [comment]
Pushed a new version.

File: Dockerfile.test [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+RUN apt update -y && apt install -y cmake gcc g++ make
+COPY . /tmp
+RUN cd /tmp && \
+    mkdir build && \
+    cd build && \
+    cmake .. -DLIBMBUS_BUILD_EXAMPLES=ON -DLIBMBUS_BUILD_TESTS=ON -DLIBMBUS_ENABLE_COVERAGE=ON && \
+    cmake --build . -j && \
+    cd .. && \
+    ./test/generate-xml.sh test/test-frames
```
Comment (lategoodbye) [inline Dockerfile.test:13]: This change looks unrelated to the pull request description. Why is it not mentioned there?

File: mbus/mbus-protocol.c [hunk 2]
```diff
@@ -2438,6 +2438,622 @@ mbus_data_error_lookup(int error)
...
+        break;
+    default:
+        snprintf(buff, sizeof(buff), "Unrecognized VIF 0xFB extension: 0x%.2x", vib->vife[0]);
+        break;
+    }
+    return buff;
+}
+
+const char *
+mbus_vib_unit_lookup_FD(mbus_value_information_block *vib)
```
Comment (lategoodbye) [inline mbus/mbus-protocol.c:2753]: Please no capital letter in function names

File: mbus/mbus-protocol.h [hunk 3]
```diff
@@ -647,6 +647,9 @@ const char *mbus_unit_prefix(int exp);
 
 const char *mbus_data_error_lookup(int error);
 
+const char *mbus_vib_unit_lookup_FB(mbus_value_information_block *vib);
+const char *mbus_vib_unit_lookup_FD(mbus_value_information_block *vib);
```
Comment (lategoodbye) [inline mbus/mbus-protocol.h:651]: Why do you think this should be available to public?

File: test/test-frames/EMU_EMU-Professional-375-M-Bus.xml [hunk 4]
```diff
@@ -241,7 +241,7 @@
     <DataRecord id="30">
         <Function>Instantaneous value</Function>
         <StorageNumber>0</StorageNumber>
-        <Unit>Unrecognized VIF extension: 0x60</Unit>
+        <Unit>Reset counter</Unit>
```
Comment (lategoodbye) [inline test/test-frames/EMU_EMU-Professional-375-M-Bus.xml:244]: I prefer to have updates to the test frames as a separate commit.

Comment (fredrik-sk) [reply test/test-frames/EMU_EMU-Professional-375-M-Bus.xml:244]: I am used to have automatic tests being executed, so I did update the code and the tests at the same time.
Do you insist on splitting it into separate commits?

File: mbus/mbus-protocol.h (hunk 3 above)
Comment (fredrik-sk) [reply mbus/mbus-protocol.h:651]: I cannot say I looked at this for very long time.
I am use to always write protypes, and I did not see a mbus-protocol-private.h where I would have put prototypes that should be considered private.

These does not have to be public. How do you normally do with such functions?

File: mbus/mbus-protocol.c (hunk 2 above)
Comment (fredrik-sk) [reply mbus/mbus-protocol.c:2753]: ok, will update

File: test/test-frames/EMU_EMU-Professional-375-M-Bus.xml (hunk 4 above)
Comment (lategoodbye) [reply test/test-frames/EMU_EMU-Professional-375-M-Bus.xml:244]: Does it mean #165 enforces this?

I didn't come a final decision about this, but i'm concerned about possible corner cases.

File: Dockerfile.test (hunk 1 above)
Comment (fredrik-sk) [reply Dockerfile.test:13]: I'm quite new to github.
So the text in the PR is not that good of a description.

This commit might fit better in the other PR (#165), it was useful for me during development, to see that the tests are actually ok.

File: test/test-frames/EMU_EMU-Professional-375-M-Bus.xml (hunk 4 above)
Comment (fredrik-sk) [reply test/test-frames/EMU_EMU-Professional-375-M-Bus.xml:244]: Yes, #165 would enforce that you have to update the tests when you change the code.

Can you describe a corner case?

I'm quite new to all this, there could be some magic string that you can put in the commit message in order to skip running the CI jobs.

My concern is that I don't want to try to do a change, and find out that the tests are not in sync with the code. ("It was broken when I got it")

Comment (lategoodbye) [reply test/test-frames/EMU_EMU-Professional-375-M-Bus.xml:244]: I'm not against automatic tests. I'm only concerned that the tests apply on commit level and not on pull request level.

Unfortunately i don't have a specific corner case in mind. On a negative side this could lead to the situation that contributors tend to put everything in a single commit.

So at least for this pull request, i'm fine with it.

File: mbus/mbus-protocol.h (hunk 3 above)
Comment (fredrik-sk) [reply mbus/mbus-protocol.h:651]: In my opinion, the functions below should not be public, so I would argue that I have been following the current standard of the code.

```
const char *mbus_vib_unit_lookup(mbus_value_information_block *vib);
const char *mbus_vif_unit_lookup(unsigned char vif);
```

**Comment by fredrik-sk** [comment]
@lategoodbye I would like input on my questions? I would like this to be done and ready ASAP.

File: mbus/mbus-protocol.h (hunk 3 above)
Comment (lategoodbye) [reply mbus/mbus-protocol.h:651]: Since these functions are only necessary for this file declare it static.

File: Dockerfile.test (hunk 1 above)
Comment (lategoodbye) [reply Dockerfile.test:13]: Okay, please move it.

**Comment by lategoodbye** [comment]
> @lategoodbye I would like input on my questions? I would like this to be done and ready ASAP.

Please point me to an question which isn't answered here.

File: test/test-frames/EMU_EMU-Professional-375-M-Bus.xml (hunk 4 above)
Comment (fredrik-sk) [reply test/test-frames/EMU_EMU-Professional-375-M-Bus.xml:244]: I'm not sure how github works. I think that github run tests on the latest commit (on a branch or pull request).

Once you are actually uesd to it, it seems natural, when I do see a code update, I expect to find corresponding updates in the tests. It is very seldom that we do change the xml-format an will have to update every test-file.

**Comment by fredrik-sk** [comment]
> 
> 
> > @lategoodbye I would like input on my questions? I would like this to be done and ready ASAP.
> 
> Please point me to an question which isn't answered here.

Sorry, but I think I reloaded and did not see any replys.

File: mbus/mbus-protocol.c (hunk 2 above)
Comment (fredrik-sk) [reply mbus/mbus-protocol.c]: Done

File: mbus/mbus-protocol.c [hunk 5]
```diff
@@ -2438,6 +2443,356 @@ mbus_data_error_lookup(int error)
...
+        case 0:
+            return "second(s)";
+        case 1:
+            return "minute(s)";
+        case 2:
+            return "hour(s)";
+        case 3:
+            return "day(s)";
+    }
+    return "<out-of-range>";
```
Comment (lategoodbye) [inline mbus/mbus-protocol.c:2460]: Please no xml style, this will break

File: mbus/mbus-protocol.c [hunk 6]
```diff
@@ -18,6 +18,11 @@
 static int parse_debug = 0, debug = 0;
 static char error_str[512];
 
+static const char *mbus_unit_duration_nn(int nn);
+static const char *mbus_unit_duration_pp(int pp);
+static const char *mbus_vib_unit_lookup_fb(mbus_value_information_block *vib);
+static const char *mbus_vib_unit_lookup_fd(mbus_value_information_block *vib);
```
Comment (lategoodbye) [inline mbus/mbus-protocol.c:24]: Please avoid forward declaration

File: mbus/mbus-protocol.h [hunk 7]
```diff
@@ -647,6 +647,9 @@ const char *mbus_unit_prefix(int exp);
...
+
```
Comment (lategoodbye) [inline mbus/mbus-protocol.h:652]: Looks like you missed them

Comment (fredrik-sk) [reply mbus/mbus-protocol.h:652]: Oops

File: mbus/mbus-protocol.c (hunk 6 above)
Comment (fredrik-sk) [reply mbus/mbus-protocol.c:24]: As I said before, I'm used to always have to have prototypes / forward declaration. Without this, the order in the file becomes important.
But if you hav that as a rule, I will comply

Comment (lategoodbye) [reply mbus/mbus-protocol.c:24]: Yes, please change this

File: mbus/mbus-protocol.c (hunk 5 above)
Comment (fredrik-sk) [reply mbus/mbus-protocol.c]: done

File: mbus/mbus-protocol.h (hunk 7 above)
Comment (fredrik-sk) [reply mbus/mbus-protocol.h]: Done

File: mbus/mbus-protocol.c (hunk 6 above)
Comment (fredrik-sk) [reply mbus/mbus-protocol.c]: Done. (removed)

My misstakes could serve as the first step towards compiling a coding standard for your project :-)
